designs/2022-eslint-to-esm/README.md

```diff
@@ -0,0 +1,212 @@
+- Repo: eslint/eslint
+- Start Date: (2022-02-28)
+- RFC PR: https://github.com/eslint/rfcs/pull/88
+- Authors: aladdin-add(weiran.zsd@outlook.com)
+
+# convert eslint codebase to esm
+
+## Summary
+
+<!-- One-paragraph explanation of the feature. -->
+Convert the ESLint codebase to ESM and publish the ESM version.
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+The Node.js community has been written commonjs for a long time, the same goes for ESLint. But with ESM introduced in ES6 and officially supported by Node.js v12, more and more projects are moving to ESM. Not long ago, we had converted a few repos under the ESLint organization to ESM:
+* @eslint/eslintrc
+* espree
+* eslint-scope
+* eslint-visitor-keys
+* @eslint/create-config
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+There are two main ways to achieve this:
+1. esm-only: write esm, publish esm
+```js
+{
+    "type": "module",
+    "exports": {
+        "./package.json": "./package.json",
+        ".": "./lib/api.js",
+        "./use-at-your-own-risk": "./lib/unsupported-api.js"
+    }
+}
+```
+2. dual-mode: compile esm to cjs, publish esm+cjs
+```js
+    "type": "module",
```

Comment:
```suggestion
```
i've made this point elsewhere, but there's no benefit to type module - it just makes `.js` files be ESM. ESM files should be `.mjs`, and then this aesthetics-only escape hatch won't be needed.

Reply:
we had been using "type": "module" in other eslint repos. I'd be supportive to change all others if want `.mjs` - for consistency.  cc @nzakas @mdjermanovic